### PR TITLE
fix mtest regression

### LINKED ISSUE #10563: testlog.txt etc don't contain output of tap tests anymore
Starting with 702f3fe15 meson-logs/testlog.txt don't contain output from tap test failures anymore. And not just when the test succeed, even when the tap test failed. @trhd I assume this was not intentional?

Reproducer:
```
project('repro', ['c'])

test('ex', find_program('echo'),
     args: ['ok 1 - frak
# diag
not ok 2 - blarg
1..2
'],
     protocol: 'tap',
    )
```
with 0.62 that'll include 
```
----------------------------------- output -----------------------------------
ok 1 - frak
# diag
not ok 2 - blarg
1..2
```
in the log, but with 702f3fe15 that's not included anymore.

Obviously testlog.txt (etc) need to include the test output...

### PR DESCRIPTION
**Pull request by trhd**
This is a cherry-picked commit from #9930. Fixes #10563 and #10577.

### CODE REVIEW

**Review by anarazel — COMMENTED** [review]
Looks like a pretty big change for a bugfix that needs to be backpatched - could this be done in a narrower fashion?

Would be good to add a test for the currently broken logging.

**Comment by eli-schwartz** [comment]
For fixing the stable release, having small changes that are easily vetted in scope is crucial.

File: test cases/common/253 long output/meson.build [hunk 1]
```diff
@@ -0,0 +1,5 @@
+project('long-stderr', 'c')
+
+dumper = executable('dumper', 'dumper.c')
+test('dump-test', dumper)
+test('dump-test-TAP', dumper, protocol : 'tap')
```
Comment (anarazel) [inline test cases/common/253 long output/meson.build:5]: Hm - nothing in this test verifies that the test log etc contains relevant output, correct?

**Comment by jpakkane** [comment]
Is this now good enough that it fixes all known regressions? IOW if we merge this can we do a point release and fix the remaining regressions in the next one?

**Comment by anarazel** [comment]
I don't know the code in mtest well enough to fully judge, but I can confirm that this fixes the problem. The expanded test in the last revision also increases coverage of the facility by a good bit, I think...

**Pull request merged**